### PR TITLE
Change project.buildDir in standalone `subprojects` property

### LINKED ISSUE #91018: android/app module's project buildDir not correct after subprojects block merged
If there is a plugin project name's lexicographical order before 'app', the prject ':app' buildDir is not set correctly.

It seems this happened after this PR merged https://github.com/flutter/flutter/pull/77942

## Step to produce:
```
flutter create -t plugin --platforms android,ios aaa
cd example
flutter build apk
```

There is an unexpected `build` folder under `android/app`.

image

flutter doctor -v
```
[✓] Flutter (Channel unknown, 2.6.0-12.0.pre.135, on macOS 11.6 20G165 darwin-x64, locale zh-Hans-CN)
    • Flutter version 2.6.0-12.0.pre.135 at /Users/fenglang/codes/source/flutter
    • Upstream repository unknown
    • Framework revision 3074c9e04b (28 hours ago), 2021-09-28 22:08:04 -0700
    • Engine revision bccb3a57eb
    • Dart version 2.15.0 (build 2.15.0-156.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/fenglang/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] VS Code (version 1.60.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.26.0

[!] Proxy Configuration
    • HTTP_PROXY is set
    ! NO_PROXY is not set

[✓] Connected device (1 available)
    • Chrome (web) • chrome • web-javascript • Google Chrome 94.0.4606.61

! Doctor found issues in 1 category.
```

### PR DESCRIPTION
**Pull request by littleGnAl**
If there is a plugin project whose name is before 'app' (in lexicographical order), the project ':app' buildDir is not set correctly.

This PR revert the changes made by https://github.com/flutter/flutter/pull/77942, I'm not sure it's the right way to fix this issue, but I add a test for this case to ensure it works as expected.

*List which issues are fixed by this PR. You must list at least one issue.*

Fixed https://github.com/flutter/flutter/issues/91018

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by jmagman** [comment]
@blasten

**Comment by blasten** [comment]
we need a test for this. would you be able to wire a plugin in the test that has the desired name?


The code below has an example about how to create a plugin in a test:

https://github.com/flutter/flutter/blob/34059eec2c671472ee95f5093705fb9fed0b3919/packages/flutter_tools/test/integration.shard/generated_plugin_registrant_test.dart#L76

**Comment by littleGnAl** [comment]
Hello @blasten, I have reused the `android_plugin_example_app_build_test.dart`, should I add a standalone integration.shard for this case?
https://github.com/flutter/flutter/pull/91030/files#diff-0a03b0a0c8e091b09635cd4d75bd08462084a15c4333fbc9fa95eab9bf2537e1R129-R130

File: packages/flutter_tools/test/integration.shard/android_plugin_example_app_build_test.dart [hunk 1]
```diff
@@ -107,6 +118,15 @@ android.enableR8=true''');
...
     ], workingDirectory: exampleAppDir.path);
+    expect(appBuildDir.existsSync(), isFalse);
     expect(exampleApk, exists);
+  }
+
+  test('plugin example can be built using current Flutter Gradle plugin', () async {
+    _createAndBuildPlugin('plugin_test');
+  });
+
+  test('plugin example android/app/build should not exists after flutter build', () async {
```
Comment (littleGnAl) [inline packages/flutter_tools/test/integration.shard/android_plugin_example_app_build_test.dart:129]: @blasten I have added a new test case here, should I move it to a standalone test file?

**Comment by blasten** [comment]
ah - that should work. does the test fail if you revert the changes to the template?

**Review by blasten — APPROVED** [review]

**Comment by littleGnAl** [comment]
> ah - that should work. does the test fail if you revert the changes to the template?

@blasten It seems no

**Review by blasten — CHANGES_REQUESTED** [review]
I see. Please verify the test fail without the fix

**Comment by littleGnAl** [comment]
@blasten The test fails without the fix.
```
➜  flutter_tools git:(littlegnal/fix-subprojects-builddir-not-correct-2) ../../bin/cache/dart-sdk/bin/pub run test /Users/fenglang/codes/source/flutter/packages/flutter_tools/test/integration.shard/android_plugin_example_app_build_test.dart
The top level `pub` command is deprecated. Use `dart pub` instead.
01:49 +1 -1: plugin example android/app/build should not exists after flutter build [E]
  Expected: false
    Actual: <true>

  package:test_api                                                          expect
  test/integration.shard/android_plugin_example_app_build_test.dart 72:5    main._createAndBuildPlugin
  test/integration.shard/android_plugin_example_app_build_test.dart 130:5   main.<fn>
  test/integration.shard/android_plugin_example_app_build_test.dart 129:82  main.<fn>
  test/src/common.dart 163:18                                               test.<fn>
  test/src/common.dart 159:5                                                test.<fn>

01:49 +1 -1: loading /Users/fenglang/codes/source/flutter/packages/flutter_tools/test/integration.shard/android_plugin_example_app_build_test.dart
Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
01:49 +1 -1: Some tests failed.
```
```
diff --git a/packages/flutter_tools/templates/app_shared/android-kotlin.tmpl/build.gradle b/packages/flutter_tools/templates/app_shared/android-kotlin.tmpl/build.gradle
index 24047dce5d..ed45c65885 100644
--- a/packages/flutter_tools/templates/app_shared/android-kotlin.tmpl/build.gradle
+++ b/packages/flutter_tools/templates/app_shared/android-kotlin.tmpl/build.gradle
@@ -21,8 +21,6 @@ allprojects {
 rootProject.buildDir = '../build'
 subprojects {
     project.buildDir = "${rootProject.buildDir}/${project.name}"
-}
-subprojects {
     project.evaluationDependsOn(':app')
 }
```

**Comment by littleGnAl** [comment]
@blasten I finally move the test to a standalone file to fill more test cases, pls take a look again, thanks.

**Comment by littleGnAl** [comment]
The tests fail without this fix.
```
➜  flutter_tools git:(littlegnal/fix-subprojects-builddir-not-correct-2) ✗ ../../bin/cache/dart-sdk/bin/pub run test /Users/fenglang/codes/source/flutter/packages/flutter_tools/test/integration.shard/flutter_build_android_app_project_builddir_test.dart
The top level `pub` command is deprecated. Use `dart pub` instead.
01:08 +0 -1: android/app/build should not exists after flutter build apk [E]
  Expected: not a file system entity that exists
    Actual: LocalDirectory:<LocalDirectory: '/private/var/folders/gk/3g8dn2ys3nv01yxsd0j_1cz00000gr/T/flutter_flutter_plugin_test.tUCf5b/aaa/example/android/app/build'>

  package:test_api                                                                  expect
  test/integration.shard/flutter_build_android_app_project_builddir_test.dart 55:5  main._checkBuildDir
  test/integration.shard/flutter_build_android_app_project_builddir_test.dart 68:7  main.<fn>
  test/integration.shard/flutter_build_android_app_project_builddir_test.dart 60:5  main.<fn>
  test/src/common.dart 163:18                                                       test.<fn>
  test/src/common.dart 159:5                                                        test.<fn>

01:08 +0 -1: loading /Users/fenglang/codes/source/flutter/packages/flutter_tools/test/integration.shard/flutter_build_android_app_project_builddir_test.dart
Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
01:47 +0 -2: android/app/build should not exists after flutter build appbundle [E]
  Expected: not a file system entity that exists
    Actual: LocalDirectory:<LocalDirectory: '/private/var/folders/gk/3g8dn2ys3nv01yxsd0j_1cz00000gr/T/flutter_flutter_plugin_test.D0m6sd/aaa/example/android/app/build'>

  package:test_api                                                                  expect
  test/integration.shard/flutter_build_android_app_project_builddir_test.dart 55:5  main._checkBuildDir
  test/integration.shard/flutter_build_android_app_project_builddir_test.dart 82:7  main.<fn>
  test/integration.shard/flutter_build_android_app_project_builddir_test.dart 74:5  main.<fn>
  test/src/common.dart 163:18                                                       test.<fn>
  test/src/common.dart 159:5                                                        test.<fn>

01:47 +0 -2: loading /Users/fenglang/codes/source/flutter/packages/flutter_tools/test/integration.shard/flutter_build_android_app_project_builddir_test.dart
Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
01:47 +0 -2: Some tests failed.
```

```
diff --git a/packages/flutter_tools/templates/app_shared/android-kotlin.tmpl/build.gradle b/packages/flutter_tools/templates/app_shared/android-kotlin.tmpl/build.gradle
index 24047dce5d..ed45c65885 100644
--- a/packages/flutter_tools/templates/app_shared/android-kotlin.tmpl/build.gradle
+++ b/packages/flutter_tools/templates/app_shared/android-kotlin.tmpl/build.gradle
@@ -21,8 +21,6 @@ allprojects {
 rootProject.buildDir = '../build'
 subprojects {
     project.buildDir = "${rootProject.buildDir}/${project.name}"
-}
-subprojects {
     project.evaluationDependsOn(':app')
 }
```

**Review by blasten — APPROVED** [review]
nice! LGTM

**Comment by gspencergoog** [comment]
This appears to have broken the build.  I'm going to try and revert and see if it fixes this.

**Comment by gspencergoog** [comment]
Actually, it doesn't appear to be the case: the revert PR failed too. Probably an infra issue.

**Pull request merged**